File: 3_Signalanalyse/MATLAB/Zeropadding.m
clear all
close all
clc

%% PART 1
sampl_freq = 50e6;      % sampling frequency

% signal generation
f_center = 1e6;         % center frequency
bandwidth = 0.6;        % relative bandwidth
tc = gauspuls('cutoff', f_center, bandwidth, [], -40);
t_center = -tc : 1/sampl_freq : tc;

y = gauspuls(t_center, f_center, bandwidth);
t = (0:length(y)-1)./sampl_freq;

plot(t*1e6, y)
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y))])

delta_f = sampl_freq/length(y);
f = -sampl_freq/2:delta_f:sampl_freq/2-delta_f;

figure
plot(f, abs(fftshift(fft(y))));
xlabel('Frequency {\itf} / Hz')
title(['Frequency domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y))])

% zeropadding
number_zeros = 1000;
y_padded = [y, zeros(1, number_zeros)];
t_padded = (0:length(y_padded)-1)./sampl_freq;
figure
plot(t_padded*1e6, y_padded)
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_padded))])

delta_f_padded = sampl_freq/length(y_padded);
f_padded = -sampl_freq/2:delta_f_padded:sampl_freq/2-delta_f_padded;

figure
plot(f_padded, abs(fftshift(fft(y_padded))));
xlabel('Frequency {\itf} / Hz')
title(['Frequency domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_padded))])

%% PART 2
% add signal with higher frequencies
c = 25;
disp(['Zusaätzliche Mittenfrequenz: ', num2str(f_center*c/1e6), ' MHz'])

y_high_freq = gauspuls(t_center, f_center*c, bandwidth/c);
y_add =y + y_high_freq;
figure
plot(t, y)
hold on
plot(t, y_high_freq, 'r')
hold on
plot(t, y_add, 'g')
xlabel('Time {\itt} / µs')
title(['Time domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_add))])
legend('lower freq.', 'high freq.', 'sum')

figure
plot(f, fftshift(abs(fft(y_add))))
xlabel('Frequency {\itf} / Hz')
title(['Frequency domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_add))])

y_add_padded = [y_add, zeros(1, number_zeros)];

figure
plot(f_padded, fftshift(abs(fft(y_add_padded))))
xlabel('Frequency {\itf} / Hz')
title(['Frequency domain. f_{sampl} = ', num2str(sampl_freq/1e6), ' MHz; N = ', num2str(length(y_add_padded))])

File: 3_Signalanalyse/Python/Zeropadding.py
from scipy import signal
import numpy as np
import matplotlib.pyplot as plt
plt.close("all")

# PART 1
sampl_freq = 50e6       # sampling frequency

# gaussian modulated sine
f_center = 1e6          # center frequency
bandwidth = 0.6         # relative bandwidth
n = 200                 # number of samples
t = np.linspace(0, n/sampl_freq, n)   # time

# without zeropadding
_, y = signal.gausspulse(t-t[-1]/2, fc=f_center, bw=bandwidth, retquad=True)

plt.figure()
plt.plot(t*1e6, y)
plt.xlabel('Time $t$ / µs')
plt.title('signal without zeropadding')

# fouriertransform
delta_f = sampl_freq/n
f = np.linspace(-sampl_freq/2, sampl_freq/2-delta_f, np.size(y))

plt.figure()
plt.plot(f/1e6, abs(np.fft.fftshift(np.fft.fft(y))))
plt.xlabel('Frequency $f$ / MHz')
plt.title('spectrum of signal without zeropadding')

# with zeropadding
number_zeros = 1000         # number of zeros padded
y_padded = np.concatenate((y, np.zeros(number_zeros)))      # padded signal
t_padded = np.linspace(0, np.size(y_padded)-1, np.size(y_padded))/sampl_freq

plt.figure()
plt.plot(t_padded*1e6, y_padded)
plt.xlabel('Time $t$ / µs')
plt.title('signal with zeropadding')

# fouriertransform
delta_f_padded = sampl_freq/np.size(y_padded)
f_padded = np.linspace(-sampl_freq/2, sampl_freq/2-delta_f_padded, np.size(y_padded))

plt.figure()
plt.plot(f_padded/1e6, abs(np.fft.fftshift(np.fft.fft(y_padded))))
plt.xlabel('Frequency $f$ / MHz')
plt.title('spectrum of signal without zeropadding')

# PART 2
# add higher frequencies
c = 23                  # shift of center frequency
_, y_high_freq = signal.gausspulse(t-t[-1]/2, fc=f_center*c, bw=bandwidth/c, retquad=True)

# add low and high frequency part
y_add = y+y_high_freq
plt.figure()
plt.plot(t*1e6, y, 'b', label='low frequency')
plt.plot(t*1e6, y_high_freq, 'r', label='high frequency')
plt.plot(t*1e6, y_add, 'g', label='sum')
plt.xlabel('Time $t$ / µs')
plt.legend()

plt.figure()
plt.plot(f/1e6, abs(np.fft.fftshift(np.fft.fft(y_add))))
plt.xlabel('Frequency $f$ / MHz')
plt.title('spectrum of signal without zeropadding')

# padded signal
y_add_padded = np.concatenate((y_add, np.zeros(number_zeros)))

plt.figure()
plt.plot(f_padded/1e6, abs(np.fft.fftshift(np.fft.fft(y_add_padded))))
plt.xlabel('Frequency $f$ / MHz')
plt.title('spectrum of signal with zeropadding')

plt.show()
